src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -196,27 +196,41 @@ struct TensorContract
 
 /*!
  * \ingroup TensorExpressionsGroup
+ * \brief Creates a contraction expression from a tensor expression if there are
+ * any indices to contract
+ *
+ * \details If there are no indices to contract, the input TensorExpression is
+ * simply returned. If not, a contraction expression is created for contracting
+ * one pair of upper and lower indices. If there is more than one pair of
+ * indices to contract, subsequent contraction expressions are recursively
+ * created, nesting one contraction expression inside another.
+ *
```

Comment:
Thinking about this some more, couple for loops might be the best option, actually. Because then you can also return the location of the repeated indices:
```cpp
constexpr std::pair first_repeated_index_locations =
  []() noexcept {
    constexpr auto tensor_index_list = 
      make_cpp20_array_from_list<TensorIndexList>();
    for (size_t i = 0; i < tensor_index_list.size(); ++i) {
      const size_t index_to_find =
            opposite_valence(tensor_index_list[i]);  // NOLINT
      for (size_t j = i + 1; j < tensor_index_list.size(); ++j) {
        if (index_to_find == tensor_index_list[j]) { // NOLINT
          return std::pair{i, j};
        }
      }
    }
    return std::pair{std::numeric_limits<size_t>::max(),
                     std::numeric_limits<size_t>::max()};
  }();
constexpr std::pair no_repeated_sentinel{std::numeric_limits<size_t>::max(),
                                         std::numeric_limits<size_t>::max()};
if constexpr (first_repeated_index_locations == no_repeated_sentinel) {
  return ~t;
} else {
  // ...
}
```

Reply:
Hmm I think I'm not following what the motivation for this change would be. To me, it seems like more verbose code that's less easy on a reader. Is the motivation just to avoid calling `contract` with any template parameters from `Tensor::operator()`, or is the motivation that this implementation is faster than using `tmpl::pop_front`? Trying to figure out if you're mainly being motivated by being able to call `contract` without a parameter, or a performance benefit.

In general though, I think this should work.

Reply:
The big motivation is removing more template code by replacing it with runtime code. You are replacing the what I would argue is very difficult to read:
```cpp
template <typename Seq, typename S, typename E>
struct contracted_indices_helper {
  using type = typename std::conditional<
      std::is_same<tmpl::count_if<
                       Seq, std::is_same<::tensorindex_with_opposite_valence<E>,
                                         tmpl::_1>>,
                   tmpl::size_t<1>>::value and
          std::is_same<
              tmpl::index_of<S, ::tensorindex_with_opposite_valence<E>>,
              tmpl::no_such_type_>::value,
      tmpl::push_back<S, E>, S>::type;
};

template <typename TensorIndexList>
using contracted_indices =
    tmpl::fold<TensorIndexList, tmpl::list<>,
               detail::contracted_indices_helper<tmpl::pin<TensorIndexList>,
                                                 tmpl::_state, tmpl::_element>>;
```
with the much easier for loops. You could also factor the for loops into a separate helper function. The above TMP is, I think, much harder to understand than two nested for loops.

```cpp
constexpr std::pair<size_t, size_t> first_contracted_indices_locations(
      const std::array<size_t, Rank>& tensor_index_list) noexcept {
    for (size_t i = 0; i < tensor_index_list.size(); ++i) {
      const size_t index_to_find =
            opposite_valence(tensor_index_list[i]);  // NOLINT
      for (size_t j = i + 1; j < tensor_index_list.size(); ++j) {
        if (index_to_find == tensor_index_list[j]) { // NOLINT
          return std::pair{i, j};
        }
      }
    }
    return std::pair{std::numeric_limits<size_t>::max(),
                     std::numeric_limits<size_t>::max()};
}
```
I find is much easier to read than the 12 lines of TMP.

Does that make my motivation clear? Basically, I'd like to replace the `contracted_indices` TMP code with the constexpr function because constexpr code is so much easier to read.

Reply:
Gotcha, this makes sense, and I'm in agreement. I think I was narrowly looking just at the `contract` function, but I agree. This is easier to follow than templated code. I'll update this, and thanks for the suggestion!

Reply:
@nilsdeppe Went ahead and implemented this in 8d86973 and swapped out `contracted_indices`.